base/uk.ac.stfc.isis.ibex.ui.configserver/src/uk/ac/stfc/isis/ibex/ui/configserver/CheckboxLabelProvider.java

```diff
@@ -19,28 +19,77 @@
 
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.WeakHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
 import org.eclipse.core.databinding.observable.map.IObservableMap;
 import org.eclipse.jface.viewers.ViewerCell;
 import org.eclipse.swt.SWT;
 import org.eclipse.swt.events.SelectionAdapter;
 import org.eclipse.swt.events.SelectionEvent;
+import org.eclipse.swt.events.SelectionListener;
 import org.eclipse.swt.widgets.Button;
+import org.eclipse.swt.widgets.Listener;
+import org.eclipse.swt.widgets.TypedListener;
 
+import uk.ac.stfc.isis.ibex.ui.tables.DataboundTable;
 import uk.ac.stfc.isis.ibex.ui.widgets.ButtonCellLabelProvider;
 
 /**
- * A LabelProvider that adds a checkbox to a cell in a table. 
+ * A LabelProvider that adds a check box to a cell in a table. 
  *
- * @param <T> The model to get the/set the information for the checkbox
+ * @param <T> The model to get the/set the information for the check box.
  */
 public abstract class CheckboxLabelProvider<T> extends ButtonCellLabelProvider<T> {
+    
+    /**A map between a model and a flag that says whether the selection listeners 
+     * on the check box on the row of the given model need to be updated. The 
+     * value of the flag is false, unless an item has just been added or removed
+     * from the table.*/
+    private Map<T, AtomicBoolean> checkboxListenerUpdateFlags = new WeakHashMap<>();
+    
+    /**The data bound table that owns this label provider.*/
+    private final DataboundTable<T> databoundTable;
+    
+    /**
+     * A selection listener that binds models to check boxes.
+     */
+    public class CheckboxSelectionAdapter extends SelectionAdapter {    
+        final Button checkbox;
+        final T model;
+        
+        /**
+         * Makes a new check box selection adapter.
+         * @param checkbox the check box a model will be bound to.
+         * @param model the model in a table item the check box will be bound to.
+         */
+        public CheckboxSelectionAdapter(Button checkbox, T model) {
+            super();
+            this.checkbox = checkbox;
+            this.model = model;
+        }
+        
+        @Override
+        public void widgetSelected(SelectionEvent e) {
+            setChecked(model, checkbox.getSelection());
+            checkbox.setText(stringFromRow(model));
+        }
+    }
 	
 	/**
 	 * The default constructor for the CheckboxLabelProvider.
-	 * @param stateProperties The properties that this label provider should be observing
+	 * @param stateProperties The properties that this label provider should be observing.
+	 * @param table The data bound table that owns this label provider.
 	 */
-	public CheckboxLabelProvider(IObservableMap stateProperties) {
+	public CheckboxLabelProvider(IObservableMap stateProperties, DataboundTable<T> table) {
```

Comment:
The observable map here contains information about the models in it's keys. So rather than putting the new `changeListener` everywhere you can do something here like:
```
stateProperties.addMapChangeListener( new IMapChangeListener() {
	@Override
	public void handleMapChange(MapChangeEvent event) {
		updateCheckboxListenerUpdateFlags(event.diff.getAddedKeys());
	}
});
```

Reply:
Done.